tiling.js

```diff
@@ -1221,26 +1208,53 @@ export class Space extends Array {
 
             // Ensure if we change workspaces and then monitors,
             // that the new workspace stays active on the starting monitor.
-            space.activateWithFocus(space.selectedWindow, false, true);
+            if (space.selectedWindow) {
+                space.activateWithFocus(space.selectedWindow, false, true);
+            }
 
             const newMonitor = Main.layoutManager.monitors[i];
             const newSpace = spaces.monitors.get(newMonitor);
+
+            newSpace.activate(false, false);
+            Navigator.finishNavigation();
+
+            // New monitor is empty, just move the mouse there
+            if (newSpace.length === 0) {
+                Utils.warpPointerToMonitor(newMonitor);
+                return;
+            }
+
+            Navigator.getNavigator().showMinimap(newSpace);
+
             const visibleColumns = newSpace.filter((column) => newSpace.isVisible(column[0]));
             if (visibleColumns.length === 0) {
                 return;
             }
 
-            const newColumn = dir === Meta.DisplayDirection.LEFT
-                ? visibleColumns[visibleColumns.length - 1]
-                : visibleColumns[0];
-            const newRow = Math.min(row, newColumn.length - 1);
-            const newWindow = newColumn[newRow];
+            const newIndex = dir === Meta.DisplayDirection.LEFT ? visibleColumns.length - 1 : 0;
+            const newColumn = visibleColumns[newIndex];
+            const newWindow = sortWindows(newSpace, newColumn)[newColumn.length - 1];
 
-            newSpace.activate(false, false);
-            Navigator.finishNavigation();
-            Navigator.getNavigator().showMinimap(newSpace);
             ensureViewport(newWindow, newSpace);
-            return
+        };
+
+        let index = space.selectedIndex();
+        if (index === -1) {
+            switchMonitor();
+            return;
+        }
+        let row = space[index].indexOf(space.selectedWindow);
+
+        switch (direction) {
+        case Meta.MotionDirection.RIGHT:
+            index++;
+            break;
+        case Meta.MotionDirection.LEFT:
+            index--;
+        }
+        if (index < 0 || index >= space.length) {
+            switchMonitor();
+            return;
         }
 
         let column = space[index];
```

Comment:
TODO: apply to switching vertically, too. So far I was just focused on horizontally because that's what I use; I don't have any monitors laid out vertically.